bin/driver.ml

```diff
@@ -36,140 +22,183 @@ let typecheck_only
-      Env.String.iter (fun name var -> Debug.print (string_of_int var ^ " -> " ^ name ^ " :: " ^
-        Types.string_of_datatype (Env.String.lookup Lib.typing_env.Types.var_env name ) )) Lib.nenv));
-
-    let load_prelude_inner () =
-      let open Loader in
-      let source =
-        (die_on_exception_unless_interacting false
-          (Loader.load_file (Lib.nenv, Lib.typing_env)) (val_of (Settings.get prelude_file)))
+    let interactive : string -> Context.t -> Sugartypes.sentence Loader.result
+      = fun ps1 context ->
+      let program, pos_context =
+        Parse.Readline.parse ps1
       in
-      let (nenv, tyenv) = source.envs in
-      let (globals, _, _) = source.program in
-
-      let tenv = (Var.varify_env (Lib.nenv, Lib.typing_env.Types.var_env)) in
+      let context' = Context.({ context with source_code = pos_context }) in
+      Loader.({ program_ = program; context = context' })
+  end
+
+  module Desugar = struct
+    let run : Sugartypes.program Loader.result -> Sugartypes.program Frontend.result
+      = fun Loader.({ context; program_ }) ->
+      Frontend.program context program_
+
+    let interactive : Sugartypes.sentence Loader.result -> Sugartypes.sentence Frontend.result
+      = fun Loader.({ context; program_ }) ->
+      Frontend.interactive context program_
+  end
+
+  module Compile = struct
+    module IR = struct
+      let run : Sugartypes.program Frontend.result -> Sugartoir.result
+        = fun Frontend.({ context; datatype; program }) ->
+        Sugartoir.program context datatype program
+    end
+  end
+
+  module Transform = struct
+    let run : Sugartoir.result -> Backend.result
+      = fun Sugartoir.({ context; datatype; globals; program }) ->
+      let program' =
+        Ir.with_bindings globals program (* TODO(dhil): This looks silly, and I think it may be wrong... *)
+      in
+      Backend.program context datatype program'
+  end
 
-      let globals = Backend.transform_prelude tenv globals in
+  module Evaluate = struct
+    module Eval = Evalir.Eval(Webserver)
 
-      let valenv = Eval.run_defs Value.Env.empty globals in
-      let envs =
-        (valenv,
-        Env.String.extend Lib.nenv nenv,
-        Types.extend_typing_environment Lib.typing_env tyenv)
+    let run : Backend.result -> (Context.t * Types.datatype * Value.t)
+      = fun Backend.({ context; program; datatype }) ->
+      let valenv = Context.value_environment context in
+      let (valenv', v) =
+        lazy (Eval.run_program valenv program) |>measure_as<| "run_program"
       in
-        globals, envs
-   in
-   die_on_exception load_prelude_inner ()
+      (Context.({ context with value_environment = valenv' }), datatype, v)
+  end
+
+  module Interactive = struct
+    type result =
+      { context: Context.t;
+        sentence: [ `Definitions of Ir.binding list
+                  | `Expression of Types.datatype * Value.t
+                  | `Directive of string * string list ] }
```

Comment:
Can we avoid a polymorphic variant here?

Reply:
Yes, easily. I mostly left whatever I could unchanged.